crates/control/src/search_suggestor.rs

```diff
@@ -0,0 +1,149 @@
+                .round() as usize,
+        );
+        let heatmap = Heatmap {
+            map: DMatrix::from_element(heatmap_length, heatmap_width, 0.0),
+            field_dimensions: context.field_dimensions.clone(),
+            cells_per_meter: context.search_suggestor_configuration.cells_per_meter,
+        };
+        Ok(Self { heatmap })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.ground_to_field.copied(),
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.map.iamax_full();
+        let mut suggested_search_position: Option<Point2<Field>> = None;
+
+        if self.heatmap.map[maximum_heat_heatmap_position]
+            > context.search_suggestor_configuration.minimum_validity
+        {
+            let search_suggestion = point![
+                ((maximum_heat_heatmap_position.0 as f32 + 1.0 / 2.0)
+                    / context.search_suggestor_configuration.cells_per_meter
+                    - context.field_dimensions.length / 2.0),
+                ((maximum_heat_heatmap_position.1 as f32 + 1.0 / 2.0)
+                    / context.search_suggestor_configuration.cells_per_meter
+                    - context.field_dimensions.width / 2.0)
+            ];
+            suggested_search_position = Some(search_suggestion);
+        }
+        context
+            .heatmap
+            .fill_if_subscribed(|| self.heatmap.map.clone());
+
+        Ok(MainOutputs {
+            suggested_search_position: suggested_search_position.into(),
+        })
+    }
+
+    fn update_heatmap(
+        &mut self,
+        ball_position: Option<&BallPosition<Ground>>,
+        invalid_ball_positions: &Vec<HypotheticalBallPosition<Ground>>,
+        ground_to_field: Option<Isometry2<Ground, Field>>,
+        heatmap_decay_factor: f32,
+    ) {
+        if let Some(ball_position) = ball_position {
+            if let Some(ground_to_field) = ground_to_field {
+                self.heatmap[ground_to_field * ball_position.position] = 1.0;
+            }
+        }
+        for ball_hypothesis in invalid_ball_positions {
+            if let Some(ground_to_field) = ground_to_field {
+                let ball_hypothesis_position = ground_to_field * ball_hypothesis.position;
+                self.heatmap[ball_hypothesis_position] =
+                    (self.heatmap[ball_hypothesis_position] + ball_hypothesis.validity) / 2.0;
+            }
+        }
+        self.heatmap.map.scale_mut(1.0 - heatmap_decay_factor);
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+struct Heatmap {
+    map: DMatrix<f32>,
+    field_dimensions: FieldDimensions,
+    cells_per_meter: f32,
+}
+
+impl Heatmap {
+    fn field_to_heatmap(&self, index: Point2<Field>) -> (usize, usize) {
```

Comment:
rename `index` to sth. like `field_point`?

Reply:
Done